File: host/cpp/adsb_json_client.cpp
// Receive decoded ADS-B JSON events from an E200 readsb instance.
// The board performs IIO acquisition and ADS-B decoding; this program only
// consumes the newline-delimited JSON result stream (default TCP port 8081).

#include <arpa/inet.h>
#include <cerrno>
#include <csignal>
#include <cstring>
#include <iostream>
#include <netinet/in.h>
#include <regex>
#include <stdexcept>
#include <string>
#include <sys/socket.h>
#include <unistd.h>

namespace {
volatile std::sig_atomic_t stop_requested = 0;

void on_signal(int) { stop_requested = 1; }

bool connect_to(const std::string &host, int port, int &fd) {
    fd = ::socket(AF_INET, SOCK_STREAM, 0);
    if (fd < 0) {
        return false;
    }

    sockaddr_in address{};
    address.sin_family = AF_INET;
    address.sin_port = htons(static_cast<uint16_t>(port));
    if (::inet_pton(AF_INET, host.c_str(), &address.sin_addr) != 1) {
        errno = EINVAL;
        ::close(fd);
        fd = -1;
        return false;
    }
    if (::connect(fd, reinterpret_cast<sockaddr *>(&address), sizeof(address)) < 0) {
        ::close(fd);
        fd = -1;
        return false;
    }
    return true;
}

std::string field(const std::string &json, const char *name) {
    const std::regex pattern(std::string("\\\"") + name +
                             "\\\"\\s*:\\s*(?:\\\"([^\\\"]*)\\\"|([^,}]+))");
    std::smatch match;
    if (!std::regex_search(json, match, pattern)) {
        return {};
    }
    return match[1].matched ? match[1].str() : match[2].str();
}

void print_event(const std::string &json) {
    std::string hex = field(json, "hex");
    if (hex.empty()) {
        hex = field(json, "icao");
    }
    std::string flight = field(json, "flight");
    if (flight.empty()) {
        flight = field(json, "callsign");
    }
    const std::string altitude = [&json] {
        std::string value = field(json, "alt_baro");
        return value.empty() ? field(json, "altitude") : value;
    }();
    const std::string lat = field(json, "lat");
    const std::string lon = field(json, "lon");

    std::cout << "ADS-B hex=" << (hex.empty() ? "?" : hex);
    if (!flight.empty()) {
        std::cout << " flight=" << flight;
    }
    if (!altitude.empty()) {
        std::cout << " alt=" << altitude;
    }
    if (!lat.empty() && !lon.empty()) {
        std::cout << " pos=" << lat << ',' << lon;
    }
    std::cout << '\n' << json << std::endl;
}
} // namespace

int main(int argc, char **argv) {
    const std::string host = argc > 1 ? argv[1] : "192.168.10.122";
    int port = 8081;
    try {
        if (argc > 2) {
            port = std::stoi(argv[2]);
        }
    } catch (const std::exception &) {
        std::cerr << "invalid TCP port: " << argv[2] << '\n';
        return 2;
    }
    if (port < 1 || port > 65535) {
        std::cerr << "TCP port must be between 1 and 65535\n";
        return 2;
    }

    std::signal(SIGINT, on_signal);
    std::signal(SIGTERM, on_signal);

    std::string pending;
    while (!stop_requested) {
        int fd = -1;
        if (!connect_to(host, port, fd)) {
            std::cerr << "connect " << host << ':' << port
                      << " failed: " << std::strerror(errno) << '\n';
            ::sleep(3);
            continue;
        }

        std::cerr << "connected to " << host << ':' << port << '\n';
        char buffer[8192];
        while (!stop_requested) {
            const ssize_t count = ::recv(fd, buffer, sizeof(buffer), 0);
            if (count <= 0) {
                break;
            }
            pending.append(buffer, static_cast<size_t>(count));
            size_t end = 0;
            while ((end = pending.find('\n')) != std::string::npos) {
                std::string line = pending.substr(0, end);
                pending.erase(0, end + 1);
                if (!line.empty() && line.front() == '{') {
                    print_event(line);
                }
            }
        }
        ::close(fd);
        if (!stop_requested) {
            std::cerr << "connection closed; reconnecting in 3 seconds\n";
            ::sleep(3);
        }
    }
    return 0;
}

File: host/python/adsb_json_client.py
#!/usr/bin/env python3
"""Receive decoded ADS-B JSON events from an E200 readsb instance.

The E200 performs IIO acquisition and Mode-S/ADS-B decoding on-board. This
client consumes the decoded newline-delimited JSON stream (TCP port 8081 by
default); it does not access IQ samples and does not require libiio.
"""

from __future__ import annotations

import argparse
import json
import socket
import sys
import time
from pathlib import Path
from typing import Any, TextIO


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(
        description="Receive decoded ADS-B results from an E200"
    )
    parser.add_argument("--host", default="192.168.10.122", help="E200 IP address")
    parser.add_argument("--port", type=int, default=8081, help="readsb JSON TCP port")
    parser.add_argument(
        "--output", type=Path, help="append complete JSON events to this JSONL file"
    )
    parser.add_argument(
        "--raw", action="store_true", help="print each complete JSON event"
    )
    parser.add_argument(
        "--once", action="store_true", help="exit after the first valid JSON event"
    )
    parser.add_argument(
        "--reconnect",
        type=float,
        default=3.0,
        help="seconds between reconnect attempts",
    )
    args = parser.parse_args()
    if not 1 <= args.port <= 65535:
        parser.error("--port must be between 1 and 65535")
    return args


def print_event(event: Any, raw: bool) -> None:
    """Print either the full event or a compact aircraft summary."""
    if raw:
        print(json.dumps(event, ensure_ascii=False, separators=(",", ":")), flush=True)
        return

    if not isinstance(event, dict):
        print("ADS-B " + repr(event), flush=True)
        return

    identity = event.get("hex") or event.get("icao") or event.get("address") or "?"
    flight = event.get("flight") or event.get("callsign") or ""
    altitude = event.get("alt_baro", event.get("altitude"))
    lat = event.get("lat")
    lon = event.get("lon")
    details = [f"hex={identity}"]
    if flight:
        details.append(f"flight={str(flight).strip()}")
    if altitude is not None:
        details.append(f"alt={altitude}")
    if lat is not None and lon is not None:
        details.append(f"pos={lat},{lon}")
    print("ADS-B " + " ".join(details), flush=True)


def receive(args: argparse.Namespace, output: TextIO | None) -> int:
    print(f"connecting to {args.host}:{args.port} ...", file=sys.stderr, flush=True)
    with socket.create_connection((args.host, args.port), timeout=10) as sock:
        sock.settimeout(None)
        print(f"connected to {args.host}:{args.port}", file=sys.stderr, flush=True)
        with sock.makefile("r", encoding="utf-8", errors="replace") as stream:
            for line in stream:
                line = line.strip()
                if not line:
                    continue
                try:
                    event = json.loads(line)
                except json.JSONDecodeError:
                    print(
                        f"warning: ignoring non-JSON line: {line[:160]}",
                        file=sys.stderr,
                    )
                    continue

                print_event(event, args.raw)
                if output:
                    output.write(
                        json.dumps(event, ensure_ascii=False, separators=(",", ":"))
                        + "\n"
                    )
                    output.flush()
                if args.once:
                    return 0
    return 1


def run(args: argparse.Namespace) -> int:
    output = args.output.open("a", encoding="utf-8") if args.output else None
    try:
        while True:
            try:
                result = receive(args, output)
                if result == 0:
                    return 0
                print("connection closed by E200", file=sys.stderr)
                if args.once:
                    return 2
            except OSError as exc:
                print(f"connection failed: {exc}", file=sys.stderr)
                if args.once:
                    return 2
            time.sleep(max(0.1, args.reconnect))
    except KeyboardInterrupt:
        print("\nstopped", file=sys.stderr)
        return 0
    finally:
        if output:
            output.close()


if __name__ == "__main__":
    raise SystemExit(run(parse_args()))
